### PR TITLE
Feature/asab library

### PR DESCRIPTION
**Pull request by bochkver**

### CODE REVIEW

File: asab/library/providers/zookeeper.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+from .abc import LibraryProviderABC
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+
+class ZooKeeperLibraryProvider(LibraryProviderABC):
```
Comment (PremyslCerny) [inline asab/library/providers/zookeeper.py:18]: @bochkver Make sure you are synchronized with Mithun here. The provider will be used by all services and based on Kazoo.

File: asab/library/providers/filesystem.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+from .abc import LibraryProviderABC
+
+
+class FileSystemLibraryProvider(LibraryProviderABC):
+
+
+	def __init__(self, app, splang_instance, path):
+		super().__init__(app, splang_instance, path)
+
+		self.LibraryBaseDir = path
```
Comment (PremyslCerny) [inline asab/library/providers/filesystem.py:13]: @bochkver Test it with both relative and absolute paths.

File: asab/library/__init__.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+		self.service = LibraryService(app, "asab.LibraryService")
+
+
+	async def initialize(self, app):
+		L.info("Sample module initialized.")
+		app.PubSub.subscribe_all(self)
+
+
+	async def finalize(self, app):
+		L.info("Sample module finalized.")
```
Comment (PremyslCerny) [inline asab/library/__init__.py:36]: @bochkver Please make sure you remove these sample logs. :)

File: asab/library/providers/abc.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+
+	async def list(self, path):
+		"""
+		Lists declarations on the given path.
+		"""
+		pass
+
+
+	async def write(self, path, data):
```
Comment (ateska) [inline asab/library/providers/abc.py:29]: No `write` in the ASAB, it is read-only provider.

File: asab/library/__init__.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+class Module(asab.Module):
+
+	def __init__(self, app):
+		super().__init__(app)
+		self.service = LibraryService(app, "asab.LibraryService")
+
+
+	async def initialize(self, app):
+		app.PubSub.subscribe_all(self)
```
Comment (ateska) [inline asab/library/__init__.py:29]: This is not needed - there is no subscription.

File: asab/library/providers/zookeeper.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		self.BasePath = url_pieces.path
+		if self.BasePath.endswith("/"):
+			self.BasePath = self.BasePath[:-1]
+
+		# Initialize ZooKeeper client
+		self.ZookeeperContainer = asab.zookeeper.container.ZooKeeperContainer(
+			self.App,
+			config_section_name='',
+			z_path=path
+		)
```
Comment (ateska) [inline asab/library/providers/zookeeper.py:33]: Please rebase to https://github.com/TeskaLabs/asab/pull/213 and test that.
Here you need to use `ZooKeeperService.build_container(...)`, don't construct the container on your own.

File: asab/library/providers/zookeeper.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+
+	async def initialize(self, app):
+		# Start the client
+		await self.ZookeeperContainer.initialize(app)
```
Comment (ateska) [inline asab/library/providers/zookeeper.py:38]: You don't need to do this explicitly.

**Review by ateska — CHANGES_REQUESTED** [review]

File: asab/library/providers/__init__.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+from .abc import LibraryProviderABC
+from .filesystem import FileSystemLibraryProvider
+from .zookeeper import ZooKeeperLibraryProvider
+
+__all__ = [
+	"LibraryProviderABC",
+	"FileSystemLibraryProvider",
+	"ZooKeeperLibraryProvider",
```
Comment (ateska) [inline asab/library/providers/__init__.py:8]: You don't need to export anything here.

File: asab/library/service.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+L = logging.getLogger(__name__)
+
+#
+
+
+class LibraryService(asab.Service):
+
+	def __init__(self, app, service_name):
+		super().__init__(app, service_name)
+		app.PubSub.subscribe_all(self)
```
Comment (ateska) [inline asab/library/service.py:15]: You don't need to subscribe b/c there is no PubSub annotation used here.

**Review by ateska — CHANGES_REQUESTED** [review]

**Review by ateska — CHANGES_REQUESTED** [review]

**Review by ateska — APPROVED** [review]

**Pull request merged**